Create an endpoint for accessing promoted content

linc opened the pull request:
* Add `Promoted` endpoint on `DiscussionsController` to access `PromotedContentModule`.
* Filter additional user info from API calls that should be PersonalInfo by default. It would otherwise be exposed by this endpoint.
* Refactor PromotedContentModule to allow controller access  via `Load` & `Validate` methods.
* Make the `SelectBy*` methods all work the same if only `selection` value is passed as `$Parameters`.
* Fix array index issue in `Prepare`.
* Fix passing ID array in `SelectByCategory`.
* Fix misc coding standards & docs throughout `PromotedContentModule`.
* Fix caching of multiple RoleIDs in selection (array to string conversion error).
* Allow Ranks & Categories to be selected by Name.
* Remove date keying of `Content` array.

tburry commented on applications/vanilla/modules/class.promotedcontentmodule.php:
```diff
@@ -89,10 +166,11 @@ public function GetData() {
     *
-    * @param array $Parameters
-    * @return boolean
+    * @param mixed $Parameters
```
FYI: You'd want to do `@param int|array $Parameters` instead of mixed.

tburry commented on applications/vanilla/views/discussions/promoted.php:
```diff
@@ -0,0 +1 @@
+<?php if (!defined('APPLICATION')) exit();
```
I'd go ahead and write the module out in this view instead of nothing.

linc commented:
Implemented line comments above. I also discovered an event name conflict in the process - using `AfterBody`, which is the Debugger hook - so I swapped that too. Hoping that's not used in some obscure place but there's no way around it and it's really bad.

tburry commented:
This is really hard to test without any documentation or examples.

tburry commented:
Consider the following path: `/discussions/promoted?selector=role&selection=Administrator`

When I run this with debug on I get data just fine, but there are a lot of notices in the page. I'd rather those be cleaned up before merge.

tburry commented:
I'm also noticing that for the json endpoint the array is keyed by date which is very odd for non-php consumers. This should just be a regular javascript array.

linc commented:
I fixed the above issues, and added/fixed the ability to select ranks & categories by name. 

I've also blocked the ability for multiple ranks or categories to be passed for now. Role already allowed multiple (on purpose) and that could be in use, so I've left it be.
